Correct and improve dev-guide section on fixing graphics-tests. (#3683)
SciTools/iris: docs/iris/src/developers_guide/graphics_tests.rst

match against a stored reference.
A "graphics test" is any test which employs this.

At present, such tests include the testing for modules `iris.tests.test_plot`
and `iris.tests.test_quickplot`, all output plots from the gallery examples
(contained in `docs/iris/example_tests`), and a few  other 'legacy' style tests

> (as described in :ref:`developer_tests`).

Reviewer: Where in `tests.rst` are the legacy style tests described? I can't find any reference to them there.

Author: Right at the top it says "There are three categories of tests within Iris: / * Unit tests / * Integration tests / * Legacy tests"

Reviewer: Indeed, however the only subsequent mention in that page is how legacy tests should be avoided. If there are any legacy style tests that use graphics tests they are not described in the page.

Author: Actually, what "legacy" means in this case is just "before we defined a separation between integration and unit tests".  That is, all the "iris/tests/test_*.py".
The implication is, we can't be bothered to remove/restructure all that, but we don't want any more like it.

I'm basically just not convinced that trying to spell that out will add anything very worthwhile ...

Reviewer: OK. There doesn't seem to be a way out of this that isn't confusing.

Graphics Testing Strategy
=========================

In the Iris Travis matrix, and over time, graphics tests must run with
multiple versions of Python, and of key dependencies such as matplotlib.
To make this manageable, the "check_graphic" test routine tests against

> multiple alternative 'correct' results.  It does this using an image "hash"

Reviewer: ```suggestion
multiple alternative 'acceptable' results.  It does this using an image "hash"
```

comparison technique which avoids storing reference images in the Iris
repository itself, to avoid space problems.

This consists of :

>  * The 'check_graphic' funciton uses a perceptual 'image hash' of the outputs

Reviewer: ```suggestion
 * The 'check_graphic' function uses a perceptual 'image hash' of the outputs
```

   (see https://github.com/JohannesBuchner/imagehash) as the basis for checking
   test results.

>  * The hashes of 'known accepted results' for each test are stored in a

Reviewer: ```suggestion
 * The hashes of known 'acceptable' results for each test are stored in a
```

   lookup dictionary, saved to the repo file
   ``lib/iris/tests/results/imagerepo.json`` .
 * An actual reference image for each hash value is stored in a *separate*
   public repository : https://github.com/SciTools/test-iris-imagehash .
 * The reference images allow human-eye assessment of whether a new output is
   judged to be 'close enough' to the older ones, or not.
 * The utility script ``iris/tests/idiff.py`` automates checking, enabling the

>    developer to easily compare proposed new 'correct' result images against the

Reviewer: ```suggestion
   developer to easily compare proposed new 'acceptable' result images against the
```

   existing accepted reference images, for each failing test.

> How to Add New Known-Valid Result Images

Reviewer: ```suggestion
How to Add New 'Acceptable' Result Images to Existing Tests
```

========================================

> When you find that a graphics test in the Iris testing suite has failed,

Reviewer: Does this process also work for a completely new test? Feels like the subtitle says so but paragraph seems to be exclusively for failures.

Author: The title is supposed to indicate that is about adding new "acceptable" result graphics for an existing testcase.  Which you will only do when the (existing) test fails for some reason.
Writing a whole *new* graphics testcase is somewhat trickier.  I decided not to explain that, after all, as we really don't want any more graphics tests unless absolutely necessary !  
It should probably only be done by people smart enough to work it out for themselves 😈

Author: By all means suggest improvements to the words so it's clearer ...
? Please ?

Reviewer: Like the logic, I'll have a go at some appropriate language. Stand by...

Reviewer: See new review

   Hit a button to either accept or reject each new result.

#. Upon acceptance of a change or a new image :

   * (A) The imagehash value of the new result image is added into the relevant
     set of 'valid result hashes' in the image result database file,
     ``tests/results/imagerepo.json`` ;

   * (B) the relevant output file in ``tests/result_image_comparison`` is
     renamed according to the image hash value, as ``<hash>.png``.
     A copy of this new PNG file must be added into the reference image
     repository at https://github.com/SciTools/test-images-scitools.
     (See below).

#. Now re-run the tests.  The 'new' result should now be recognised and the
   relevant test should pass.  However, certain tests do perform *multiple*
   graphics checks within a single test function : In those cases, any failing
   check will prevent the others from being run, so a test re-run can encounter
   further (new) graphical test failures.  Simply repeat the check-and-accept
   process until all graphical tests pass.

#. To add your changes to Iris, you need to make two pull requests :

   * (A) The first PR is made in the test-images-scitools repository, at

>      https://github.com/SciTools/test-images-scitools.  This should contain all

Reviewer: Just visited this repository and it claims to be deprecated in favour of:
https://github.com/SciTools/test-iris-imagehash

Author: :+1: Good spot !
Will fix.

     the newly-generated PNG files, which must be added to the ``images/v4``
     directory.  In your Iris repo, these files are to be found in the
     temporary results folder ``iris/tests/result_image_comparison``.

>      **Note**: this location is covered by a project ``.gitignore`` setting,

Author: Maybe just "this folder" - would that do ?

>      **Note**: this location is covered by a project ``.gitignore`` setting,

Reviewer: Providing you maintain the distinct uses of "directory" and "folder" in the paragraph, that should work. I personally would have explicitly said "the `result_image_comparison` folder is covered...", but I'm not precious about it!

>      **Note**: this location is covered by a project ``.gitignore`` setting,

Author: :+1:
